### DISCUSSION TITLE
Static html files do not have extension - can I change this somewhere?

### DISCUSSION THREAD

**Discussion by ghost** [datenstrom/community#317] [Ask a question] [answered]
Hi !

I don't know if here is the right place to ask my question.
I started using yellow a few day ago. It works great on a dynamic php server (hiawatha webserver). but I also want to export it as a static website destined to be used locally on a rudimentary server (busybox httpd).

The issue is that exported html files do not have the .html extension.

To be more specific, the index files (page.txt) are correctly outputed as index.html. Same goes for other files like 404.html, but other generic files (e.g. blablabla.txt) are outputed without extension (e.g. 'blablabla' instead of 'blablabla.html').

Most servers automatically guess that these are html files (That is why it works on my dynamic server even with a cached static folder). But my httpd server do not and fall back to default MIME 'application/octet-stream'.
 
Here is my command : `php yellow.php build && busybox httpd -p 127.0.0.1:8181 -f -h cache`

Is this lack of extension intended ? If so, does someone which lines of which files should I edit to change it ?
I looked into various system and config files, but I didn't find if by myself.

Maybe the config.ini could have a line like `HtmlExtension: .html` ?

Note : changing the httpd server settings, is not an option because I want to be able to serve the static website on different computers with the same default config.

**Comment by markseuffert** [comment, marked as answer]
Hi Xisd!

Yes, this behaviour is by design.

Datenstrom Yellow uses [clean URLs](https://en.wikipedia.org/wiki/Clean_URL). The file `content/9-about/what-we-do.txt` will be available as `http://website/about/what-we-do`. The URL will be the same as a static website and as a regular website. I am sorry to say, this behaviour can't be changed in the configuration. I understand why you would like to generate static files with the `.html` file extension, but that's something we consciously decided not to do. Here's what you can do:

* Configure your web server to send no default MIME type
* Configure your web server to send default MIME type `text/html`
* Configure MIME type `text/html` for files without an extension
* Use the built-in web server that's included in every installation

Can you [contact busybox](https://www.busybox.net/FAQ.html#helpme) developers, maybe they have a better suggestion? I guess that other static site generators run into the same problem. Testing and feedback is much appreciated.

**Comment by ghost** [comment]
Thank you for your answer

I must say that I am very happy to have found yellow, I've been looking for something like this for a while. The file structure is quite clear, the design is clean and usable, YAML and mardown are supported, and it does work without javascript enabled (that last one was the hardest to find).
I wasn't especially looking for an admin  interface but I like this one.
I'm glad to learn there is a buit-in web server, I somehow missed that info in the documentation.

Your solutions would solve the issue (except for the first one: that is what I did, without configuring anything MIME type for files without an extension defaults to  'application/octet-stream') but I was looking for something that would allow me to deploy my static website on a fresh linux installation (with busybox installed but not php-cli, and without the need of editing configuration files (especially when they require admin privileges and editing busybox httpd mimetypes does)).

I sent a mail to busybox developers, but I think their settings actually make sense.  
And more generally I think this means that while 'clean url' might present some visual benefit, it has a cost of compatibility. Hence my suggestion of making it an optional setting switchable from the config.ini

If I wanted to edit manually the functions that are creating the html files, wouldn't it be just a few variables to change ? If so, could you point me out to the right ones? From what I've seen, I think they would be in the core.php, but I don't know php well enough to do more than a few guess (that did not work so far).

**Comment by markseuffert** [comment]
Thanks for your feedback.

Here's the problem from a technical point of view. When you [build a static website](https://github.com/annaesvensson/yellow-command), the complete website is turned into static files. If the web server sends a HTTP response for a static file, it usually detects the MIME type based on the file extension. If the web server can't detect the MIME type for a static file, then there a different ways to proceed. One is to send a `Content-Type` header with a default MIME type, another is to send a  HTTP response without `Content-Type` header. The later gives the web browser a chance to detect the correct media type. 

See also [HTTP 1.0](https://tools.ietf.org/html/rfc1945#section-7.2.1),  [HTTP 1.1](https://tools.ietf.org/html/rfc2616#section-7.2.1), [Apache 2.4 documentation](http://httpd.apache.org/docs/2.4/mod/core.html#defaulttype), [MIME type for unknown data](https://stackoverflow.com/questions/1176022/unknown-file-type-mime/28652339#28652339)

In a nutshell, if your web server doesn't send a "default MIME type" it works out of the box.

Thanks for sending a mail to the busybox developers. Let's see what they answer. :smile:

**Comment by markseuffert** [comment]
Here's a possible patch for busybox. Tested on Ubuntu Linux that the static website works.

```
[PATCH 1/1] httpd: No Content-Type header if MIME type is unknown

---
 networking/httpd.c | 14 +++++++++-----
 1 file changed, 9 insertions(+), 5 deletions(-)

diff --git a/networking/httpd.c b/networking/httpd.c
index 9439e206c..a7138da94 100644
--- a/networking/httpd.c
+++ b/networking/httpd.c
@@ -1079,14 +1079,19 @@ static void send_headers(int responseNum)
 	/* ^^^ using gmtime_r() instead of gmtime() to not use static data */
 	len = sprintf(iobuf,
 			"HTTP/1.0 %d %s\r\n"
-			"Content-type: %s\r\n"
 			"Date: %s\r\n"
 			"Connection: close\r\n",
 			responseNum, responseString,
-			/* if it's error message, then it's HTML */
-			(responseNum == HTTP_OK ? found_mime_type : "text/html"),
 			date_str
 	);
+	
+	if (responseNum != HTTP_OK || *found_mime_type) {
+		/* Responding with content type, if error message or known MIME type */
+		len += sprintf(iobuf + len,
+				"Content-type: %s\r\n",
+				(responseNum != HTTP_OK ? "text/html" : found_mime_type)
+		);
+	}
 
 #if ENABLE_FEATURE_HTTPD_BASIC_AUTH
 	if (responseNum == HTTP_UNAUTHORIZED) {
@@ -1684,8 +1689,7 @@ static NOINLINE void send_file_and_exit(const char *url, int what)
 	 * (happens if you abort downloads from local httpd): */
 	signal(SIGPIPE, SIG_IGN);
 
-	/* If not found, default is "application/octet-stream" */
-	found_mime_type = "application/octet-stream";
+	found_mime_type = "";
 	suffix = strrchr(url, '.');
 	if (suffix) {
 		static const char suffixTable[] ALIGN1 =
-- 
2.13.6 (Apple Git-96)
```

**Comment by ghost** [comment]
Thank you for the explication, I didn't known about the `Content-Type` header. And thank you for writting a patch, did you submitted it to be merged in busybox git ?

As a temporary quick fix, I wanted to add a httpd.conf file in the root path of the built site. (or somewhere else with the -c option of httpd) but I don't think there is a line changing MIME for files without extension (at least I did not find one inside the template described in [httpd.c](https://git.busybox.net/busybox/tree/networking/httpd.c) file).
I also thought that a cgi based solution might work, there are some informations about that on a page I found on [Open Wrt Wiki](https://wiki.openwrt.org/doc/howto/http.httpd). but I don't really know how to do that.

> Maybe the config.ini could have a line like HtmlExtension: .html ?

Also, I'm sorry for asking again but I didn't get what you think about making the 'cean url' feature toogleable from the config.ini, even maybe the possiblity to choose the extension for outputed file the same way it is possible to set the ContentExtension for markdown files. This feature exist in some other static static generators, (ikiwiki for exemple), and I think that it is that could be a useful addition regardless of the current issue.

**Comment by markseuffert** [comment]
I am waiting for feedback. Perhaps you can submit the patch or add the feature to Datenstrom Yellow?

**Comment by ghost** [comment]
I sent  a message to the busybox mailing list  regarding this issue with a link to your post with the patch.
(no answer to the previous one yet)

>  add the feature to Datenstrom Yellow?

I would gladly do that but I'm afraid that may be beyond my skills in php programming, I can't even find which function is giving the name to the static html file. :confused:

**Comment by markseuffert** [comment]
The latest version of busybox works perfect, good team work. 😄

**Comment by ghost** [comment]
Thats good news! I wanted to test it too but I failed to compile it on Lubuntu 16.04.

**Comment by markseuffert** [comment]
Perhaps you can give the maintainer a positive feedback that it works on Ubuntu Linux?

**Comment by ghost** [comment]
> Perhaps you can give the maintainer a positive feedback that it works on Ubuntu Linux?

Done
